doc/neps/nep-0023-backwards-compatibility.rst

```diff
@@ -19,46 +19,206 @@ processes for individual cases where breaking backwards compatibility
+  place.
+- shall be mentioned in the documentation for the functionality. A
+  ``.. deprecated::`` directive can be used for this.
+
+Examples of good deprecation warnings (also note standard form of the comments
+above the warning, helps when grepping):
+
+.. code-block:: python
+
+    # NumPy 1.15.0, 2018-09-02
+    warnings.warn('np.asscalar(a) is deprecated since NumPy 1.16.0, use '
+                  'a.item() instead', DeprecationWarning, stacklevel=3)
+
+    # NumPy 1.15.0, 2018-02-10
+    warnings.warn("Importing from numpy.testing.utils is deprecated "
+                  "since 1.15.0, import from numpy.testing instead.",
+                  DeprecationWarning, stacklevel=2)
+
+    # NumPy 1.14.0, 2017-07-14
+    warnings.warn(
+        "Reading unicode strings without specifying the encoding "
+        "argument is deprecated since NumPy 1.14.0. Set the encoding, "
+        "use None for the system default.",
+        np.VisibleDeprecationWarning, stacklevel=2)
+
+.. code-block:: C
+
+        /* DEPRECATED 2020-05-13, NumPy 1.20 */
+        if (PyErr_WarnFormat(PyExc_DeprecationWarning, 1,
+                matrix_deprecation_msg, ufunc->name, "first") < 0) {
+            return NULL;
+        }
+
+Removal of deprecated functionality:
+
+- shall be done after at least 2 releases (assuming the current 6-monthly
+  release cycle; if that changes, there shall be at least 1 year between
+  deprecation and removal).
+- shall be listed in the release notes of the release where the removal happened.
+- can be done in any minor (but not bugfix) release.
+
+For backwards incompatible changes that aren't "deprecate and remove" but for
+which code will start behaving differently, a ``FutureWarning`` should be
+used. Release notes, mentioning version number and using ``stacklevel`` should
+be done in the same way as for deprecation warnings. A ``.. versionchanged::``
+directive can be used in the documentation to indicate when the behavior
+changed:
+
+.. code-block:: python
+
+    def argsort(self, axis=np._NoValue, ...):
+        """
+        Parameters
+        ----------
+        axis : int, optional
+            Axis along which to sort. If None, the default, the flattened array
+            is used.
+
+            ..  versionchanged:: 1.13.0
+                Previously, the default was documented to be -1, but that was
+                in error. At some future date, the default will change to -1, as
+                originally intended.
+                Until then, the axis should be given explicitly when
+                ``arr.ndim > 1``, to avoid a FutureWarning.
+        """
+        ...
+        warnings.warn(
+            "In the future the default for argsort will be axis=-1, not the "
+            "current None, to match its documentation and np.argsort. "
+            "Explicitly pass -1 or None to silence this warning.",
+            MaskedArrayFutureWarning, stacklevel=3)
+
+
+Decision making
+~~~~~~~~~~~~~~~
```

Comment:
shouldn't this be one heading level higher (same with the more strict policy below)?  Since it more importantly applies to the decision whether a change should be made at all.

Reply:
good point, done